### ISSUE TITLE
question about training loss

### ISSUE THREAD

**Issue by LyyyyyyyN** [QVPR/Patch-NetVLAD#54]
Hi,
I really appreciate for the work! It is helpful for me to try it with your instructions.
But I have a question about training. I try to retrain the network (with original Vgg16 backbone or change it to ResNet18, cropped it properly and maintain most of the parameters as the pretrained weights). I found that the loss went down rapidly at the first epoch and almost reach to 0.00. Would you describe the change of your loss when you train this network?  Thank you very much!

Best regards

**Comment by Tobias-Fischer** [comment]
Hi,

Have you tried training the network without any changes (with VGG and the Mapillary dataset), as described in https://github.com/QVPR/Patch-NetVLAD#training? The loss shouldn't go down to 0.0 at the end of the first epoch. Could you share the loss curve with us?

**Comment by LyyyyyyyN** [comment]
Thanks for your reply! Yes, I tried it with vgg-16 and Mapillary dataset as is described and didn't change the hyperparameters: 
![image](https://user-images.githubusercontent.com/74396952/163941075-d52214b3-d1f8-4120-a132-c27bec51f40e.png)
This is the loss of epoch 1 and 2 and it costs around 17 hours per epoch (Is it abnormal?) 
Also, I tried with ResNet-18 (only trained the last basicblock) and the loss for epoch 1 and 2 seems like:
![image](https://user-images.githubusercontent.com/74396952/163941881-67f7000e-f8d8-405c-8161-79403ad04135.png)
The loss is low and the recall is also low. The validation recall@1 is no more than 50% even if I train it to epoch 8 or epoch 9 (with ResNet). For VGG I didn't change anything and for ResNet I just changed a little in model_generic part. That confused me.
BTW, Would you please provide your email address so that I can describe more details to you? Thank you for your help and patience!

**Comment by Tobias-Fischer** [comment]
@StephenHausler you stared at these curves more than I have, any input? Are we doing the validation correctly those days? At some point we didn't implement it properly.

@LyyyyyyyN - if you have more questions, please ask here so that the community can benefit as well.

**Comment by StephenHausler** [comment]
Yea, the loss should drop a lot in the first epoch. Each epoch is very large, since there are just so many query images in the mapillary dataset - a single epoch with mapillary is worth ~100 pittsburgh epochs (an approximate guess). My intuition is that further training basically just refines the network to handle the edge-case hard negatives. At later epochs, most query images will give a loss of 0, there will only be a sparse set of queries that give a positive loss. The validation result will be almost at the maximum possible after just one epoch. 

If you want to see a more fine-grained view of the validation curve, you could try modifying msls.py to set a custom epoch size, say 10k images - basically randomly sample the full 1M set and take a random slice of 10k images per epoch. Each epoch would not be identical however, but this would be a quick way of observing the validation improvement in the early stages of training. You could also try lowering the learning rate, we haven't tried lowering it yet, but it is possible that the network could be trained better with a lower LR.

Validation of 50% seems ok, and this should get a little higher after running PCA. I don't think this is a problem, the validation set is challenging. This is just NetVLAD by itself of course, the recall@1 on val will be much higher when running Patch-NetVLAD.

17 hours per epoch is normal - there are about 1 million images per epoch.

**Comment by LyyyyyyyN** [comment]
Thank you so much for your detailed reply and suggestions!
